src/Oxpecker/ModelBinder.fs

```diff
@@ -14,299 +12,259 @@ open Microsoft.Extensions.Primitives
-                        (emptyList, None)
-                match error with
-                | Some err -> Error err
-                | None -> Ok items
-        elif isGeneric then
-            let result = parseValue genArgType rawValues culture
-            match result with
-            | Error err -> Error err
-            | Ok value ->
-                match isOption with
-                | false -> Ok value
-                | true ->
-                    match value with
-                    | null -> t.MakeNoneCase()
-                    | v -> t.MakeSomeCase(v)
-                    |> Ok
-        elif FSharpType.IsUnion t then
-            let unionName = rawValues.ToString()
-            let cases = FSharpType.GetUnionCases t
-            if String.IsNullOrWhiteSpace unionName then
-                Error $"Cannot parse an empty value to type %s{t.ToString()}."
-            else
-                cases
-                |> Array.tryFind(_.Name.Equals(unionName, StringComparison.OrdinalIgnoreCase))
-                |> function
-                    | Some case -> Ok(FSharpValue.MakeUnion(case, [||]))
-                    | None -> Error $"The value '%s{unionName}' is not a valid case for type %s{t.ToString()}."
-        else
-            let converter =
-                if t.GetTypeInfo().IsValueType then
-                    nullableGenericType.MakeGenericType([| t |])
+    open TypeShape.Core
+    open TypeShape.Core.Utils
+
+    let private firstValue (rawValues: StringValues) =
+        if rawValues.Count = 0 then null else rawValues[0]
+
+    let private unsupported ty = failwith $"Unsupported type '{ty}'."
+
+    let private error (values: StringValues) : 'T =
+        let value =
+            match values |> firstValue with
+            | null -> "null"
+            | _ -> values.ToString()
+
+        failwith $"Could not parse value '{value}' to type '{typeof<'T>}'."
+
+    let private (|UnionCase|_|) (shape: ShapeFSharpUnion<'T>) (caseName: string) =
+        let unionCaseExists caseName (case: ShapeFSharpUnionCase<'T>) =
+            String.Equals(case.CaseInfo.Name, caseName, StringComparison.OrdinalIgnoreCase)
+
+        shape.UnionCases |> Array.tryFind(unionCaseExists caseName)
+
+    let private (|RawValue|_|) (dict: IDictionary<string, StringValues>) =
+        match dict |> Seq.tryExactlyOne with
+        | Some(KeyValue("", value)) -> Some value
+        | _ -> None
+
+    let private (|RawValueQuick|) (dict: IDictionary<string, StringValues>) =
+        match dict with
+        | RawValue v -> v
+        | _ -> failwith $"Dictionary %A{dict} should contain only one value but it has %i{dict.Count}."
+
+    let private (|SimpleArray|_|) (data: IDictionary<string, StringValues>) =
+        match data with
+        | RawValue values when values.Count > 0 -> values |> Seq.map String.toDict |> Some
+        | _ -> None
+
+    let private (|ComplexArray|_|) (data: IDictionary<string, StringValues>) =
```

Comment:
Why does this function return dictionary? As far as I understand, it can return a sequence